File: python/src/bursar/repositories/billing/refund.py
from __future__ import annotations

from bursar.repositories._types import DbQuery
from bursar.repositories._utils import validate_non_empty


class BillingRefundRepository:
    """Repository for billing refund operations.

    All methods call Postgres via raw SQL queries through the query function.
    """

    def __init__(self, execute: DbQuery) -> None:
        self._execute = execute

    def upsert(
        self,
        provider: str,
        provider_refund_id: str,
        provider_payment_id: str | None,
        user_id: str | None,
        amount_minor: int,
        currency: str,
        reason: str | None,
        metadata: str | None,
    ) -> None:
        """Insert or update a billing refund record.

        Args:
            provider: The billing provider identifier.
            provider_refund_id: The provider refund ID.
            provider_payment_id: The associated provider payment ID, or None.
            user_id: The user ID, or None.
            amount_minor: The refund amount in minor currency units.
            currency: The ISO 4217 currency code.
            reason: The refund reason, or None.
            metadata: JSON metadata string, or None.
        """
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_refund_id, "provider_refund_id")
        self._execute(
            "SELECT public.upsert_billing_refund(%s, %s, %s, %s, %s, %s, %s, %s)",
            [provider, provider_refund_id, provider_payment_id, user_id, amount_minor, currency, reason, metadata],
        )

File: javascript/src/repositories/billing/refund.ts
import type { QueryFn } from "../types.js";

export class BillingRefundRepository {
  constructor(private query: QueryFn) {}

  async upsert(
    provider: string,
    providerRefundId: string,
    providerPaymentId: string | null,
    userId: string | null,
    amountMinor: number,
    currency: string,
    reason: string | null,
    metadata: string | null,
  ): Promise<void> {
    await this.query(`SELECT bursar.upsert_billing_refund($1, $2, $3, $4, $5, $6, $7, $8)`, [
      provider,
      providerRefundId,
      providerPaymentId,
      userId,
      amountMinor,
      currency,
      reason,
      metadata,
    ]);
  }
}
